File: frontend/src/utils/__tests__/patterns.test.ts
import { describe, it, expect } from 'vitest';
import {
  matchWildcard,
  parseFilterPatterns,
  serializeFilterPatterns,
  validateBasePattern,
  validatePatterns,
  validateFilterPatterns,
  matchesFilter,
  filterByPatterns,
} from '../patterns';

describe('patterns utils', () => {
  describe('matchWildcard', () => {
    it('matches exact strings case-insensitively', () => {
      expect(matchWildcard('foo', 'foo')).toBe(true);
      expect(matchWildcard('FOO', 'foo')).toBe(true);
      expect(matchWildcard('foo', 'FOO')).toBe(true);
      expect(matchWildcard('foo', 'bar')).toBe(false);
    });

    it('matches * wildcards', () => {
      expect(matchWildcard('*', 'anything')).toBe(true);
      expect(matchWildcard('kubernetes/*', 'kubernetes/kubernetes')).toBe(true);
      expect(matchWildcard('kubernetes/*', 'kubernetes/community')).toBe(true);
      expect(matchWildcard('kubernetes/*', 'golang/go')).toBe(false);
      expect(matchWildcard('*bug*', 'kind/bug/fix')).toBe(true);
      expect(matchWildcard('*bug*', 'kind/feature')).toBe(false);
    });

    it('matches ? single character wildcards', () => {
      expect(matchWildcard('size/?', 'size/s')).toBe(true);
      expect(matchWildcard('size/?', 'size/m')).toBe(true);
      expect(matchWildcard('size/?', 'size/xxl')).toBe(false);
    });
  });

  describe('parseFilterPatterns', () => {
    it('parses multi-line includes and excludes', () => {
      const input = 'kubernetes/*\n!kubernetes/steering\ngolang/*\n!golang/proposal\n  \n';
      const result = parseFilterPatterns(input);
      expect(result).toEqual({
        includes: ['kubernetes/*', 'golang/*'],
        excludes: ['kubernetes/steering', 'golang/proposal'],
      });
    });

    it('ignores empty lines and whitespace', () => {
      expect(parseFilterPatterns('')).toEqual({ includes: [], excludes: [] });
      expect(parseFilterPatterns('   \n\n  ')).toEqual({ includes: [], excludes: [] });
    });
  });

  describe('serializeFilterPatterns', () => {
    it('serializes includes and excludes to multi-line string', () => {
      const serialized = serializeFilterPatterns(
        ['kubernetes/*', 'golang/*'],
        ['kubernetes/steering', 'golang/proposal']
      );
      expect(serialized).toBe('kubernetes/*\ngolang/*\n!kubernetes/steering\n!golang/proposal');
    });

    it('handles empty or missing arrays', () => {
      expect(serializeFilterPatterns(undefined, undefined)).toBe('');
      expect(serializeFilterPatterns(['kubernetes/*'], undefined)).toBe('kubernetes/*');
      expect(serializeFilterPatterns(undefined, ['kubernetes/test'])).toBe('!kubernetes/test');
    });
  });

  describe('validateBasePattern', () => {
    it('validates standard patterns and handles ! prefix', () => {
      const validInclude = validateBasePattern('kubernetes/*', 'Test');
      expect(validInclude.valid).toBe(true);
      if (validInclude.valid) {
        expect(validInclude.cleanPattern).toBe('kubernetes/*');
        expect(validInclude.isExclude).toBe(false);
      }

      const validExclude = validateBasePattern('!kubernetes/steering', 'Test');
      expect(validExclude.valid).toBe(true);
      if (validExclude.valid) {
        expect(validExclude.cleanPattern).toBe('kubernetes/steering');
        expect(validExclude.isExclude).toBe(true);
      }
    });

    it('errors on empty pattern or empty exclude', () => {
      expect(validateBasePattern('', 'Test').error).toBe('Test pattern cannot be empty');
      expect(validateBasePattern('   ', 'Test').error).toBe('Test pattern cannot be empty');
      expect(validateBasePattern('!', 'Test').error).toBe('Test pattern cannot be empty');
      expect(validateBasePattern('!  ', 'Test').error).toBe('Test pattern cannot be empty');
    });

    it('errors on control characters or excessive length', () => {
      expect(validateBasePattern('test\x00pattern', 'Test').error).toContain('contains invalid control characters');
      expect(validateBasePattern('a'.repeat(101), 'Test').error).toContain('exceeds maximum length');
    });
  });

  describe('validatePatterns and validateFilterPatterns', () => {
    it('validates pattern arrays and strings using validator callback', () => {
      const validator = (p: string) => (p.includes('invalid') ? 'Invalid pattern' : null);

      expect(validatePatterns(['good', 'valid', ''], validator)).toBeNull();
      expect(validatePatterns(['good', 'invalid', 'valid'], validator)).toBe('Invalid pattern');

      expect(validateFilterPatterns('good\nvalid\n', validator)).toBeNull();
      expect(validateFilterPatterns('good\ninvalid\nvalid', validator)).toBe('Invalid pattern');
    });
  });

  describe('matchesFilter', () => {
    it('returns true when no filters are set', () => {
      expect(matchesFilter('anything', undefined, undefined)).toBe(true);
      expect(matchesFilter('anything', [], [])).toBe(true);
    });

    it('handles includes and excludes with exclude precedence', () => {
      const includes = ['kubernetes/*', 'golang/*'];
      const excludes = ['kubernetes/steering'];

      expect(matchesFilter('kubernetes/kubernetes', includes, excludes)).toBe(true);
      expect(matchesFilter('golang/go', includes, excludes)).toBe(true);
      expect(matchesFilter('kubernetes/steering', includes, excludes)).toBe(false);
      expect(matchesFilter('rust-lang/rust', includes, excludes)).toBe(false);
    });
  });

  describe('filterByPatterns', () => {
    it('filters items using identifier callback', () => {
      const items = [
        { id: 1, name: 'kubernetes/kubernetes' },
        { id: 2, name: 'kubernetes/steering' },
        { id: 3, name: 'golang/go' },
        { id: 4, name: 'rust-lang/rust' },
      ];

      const result = filterByPatterns(
        items,
        item => item.name,
        ['kubernetes/*', 'golang/*'],
        ['kubernetes/steering']
      );

      expect(result).toEqual([
        { id: 1, name: 'kubernetes/kubernetes' },
        { id: 3, name: 'golang/go' },
      ]);
    });
  });
});

File: backend/internal/logic/patterns_test.go
package logic

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestMatchWildcard(t *testing.T) {
	tests := []struct {
		pattern string
		text    string
		match   bool
	}{
		{"*", "anything", true},
		{"*/*", "owner/repo", true},
		{"kubernetes/*", "kubernetes/kubernetes", true},
		{"kubernetes/*", "golang/go", false},
		{"size/?", "size/s", true},
		{"size/?", "size/medium", false},
		{"*bug*", "kind/bug/fix", true},
		{"exact-match", "exact-match", true},
		{"EXACT-MATCH", "exact-match", true},
		{"exact-match", "EXACT-MATCH", true},
	}

	for _, tt := range tests {
		t.Run(tt.pattern+"_"+tt.text, func(t *testing.T) {
			assert.Equal(t, tt.match, MatchWildcard(tt.pattern, tt.text))
		})
	}
}

func TestMatchesFilter(t *testing.T) {
	// Empty patterns allow everything
	assert.True(t, MatchesFilter("kubernetes/kubernetes", nil, nil))
	assert.True(t, MatchesFilter("kubernetes/kubernetes", []string{}, []string{}))

	// Include patterns
	assert.True(t, MatchesFilter("kubernetes/kubernetes", []string{"kubernetes/*"}, nil))
	assert.False(t, MatchesFilter("golang/go", []string{"kubernetes/*"}, nil))

	// Exclude patterns
	assert.True(t, MatchesFilter("kubernetes/kubernetes", nil, []string{"kubernetes/test-infra"}))
	assert.False(t, MatchesFilter("kubernetes/test-infra", nil, []string{"kubernetes/test-infra"}))

	// Combined include and exclude (exclude wins)
	includes := []string{"kubernetes/*", "golang/*"}
	excludes := []string{"kubernetes/steering", "golang/proposal"}

	assert.True(t, MatchesFilter("kubernetes/kubernetes", includes, excludes))
	assert.True(t, MatchesFilter("golang/go", includes, excludes))
	assert.False(t, MatchesFilter("kubernetes/steering", includes, excludes))
	assert.False(t, MatchesFilter("golang/proposal", includes, excludes))
	assert.False(t, MatchesFilter("rust-lang/rust", includes, excludes))
}

func TestValidatePatternBase(t *testing.T) {
	// Valid patterns
	clean, err := ValidatePatternBase("kubernetes/*", "test")
	require.NoError(t, err)
	assert.Equal(t, "kubernetes/*", clean)

	// Handles ! prefix
	clean, err = ValidatePatternBase("!kubernetes/steering", "test")
	require.NoError(t, err)
	assert.Equal(t, "kubernetes/steering", clean)

	// Empty errors
	_, err = ValidatePatternBase("", "test")
	require.Error(t, err)
	assert.Contains(t, err.Error(), "test pattern cannot be empty")

	_, err = ValidatePatternBase("!  ", "test")
	require.Error(t, err)

	// Control characters error
	_, err = ValidatePatternBase("invalid\x00pattern", "test")
	require.Error(t, err)
	assert.Contains(t, err.Error(), "contains invalid control characters")

	// Max length
	longStr := strings.Repeat("a", MaxPatternLength+5)
	_, err = ValidatePatternBase(longStr, "test")
	require.Error(t, err)
	assert.Contains(t, err.Error(), "exceeds maximum length")
}

func TestValidatePatterns(t *testing.T) {
	err := ValidatePatterns([]string{"a", "b", ""}, func(p string) error {
		return nil
	})
	require.NoError(t, err)

	err = ValidatePatterns([]string{"a", "invalid"}, func(p string) error {
		if p == "invalid" {
			return assert.AnError
		}
		return nil
	})
	require.ErrorIs(t, err, assert.AnError)
}
